### ISSUE TITLE
[sw,meson] Infinite ninja build loop

### ISSUE THREAD

**Issue by GregAC** [lowRISC/opentitan#2428]
I quite commonly encounter a situation where a `ninja -C ./built-out` causes some kind of infinite loop. This seems to occur when I run `ninja -C ./build-out` having had a successful build previously, changed some file and then run again. Here is an example of the output:

```Found ninja-1.8.2 at /usr/bin/ninja
WARNING: Cross file does not specify strip binary, result will not be stripped.
WARNING: Cross file does not specify strip binary, result will not be stripped.
[0/1] Regenerating build files.
The Meson build system
Version: 0.53.2
Source dir: /home/greg/work/opentitan_area/opentitan
Build dir: /home/greg/work/opentitan_area/opentitan/build-out
Build type: cross build
Project name: opentitan
Project version: 0.1
C compiler for the build machine: cc (gcc 7.5.0 "cc (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0")
C linker for the build machine: cc ld.bfd 2.30
C++ compiler for the build machine: c++ (gcc 7.5.0 "c++ (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0")
C++ linker for the build machine: c++ ld.bfd 2.30
C compiler for the host machine: /tools/riscv/bin/riscv32-unknown-elf-gcc (gcc 9.2.0 "riscv32-unknown-elf-gcc (GCC) 9.2.0")
C linker for the host machine: /tools/riscv/bin/riscv32-unknown-elf-gcc ld.bfd 2.32
C++ compiler for the host machine: /tools/riscv/bin/riscv32-unknown-elf-g++ (gcc 9.2.0 "riscv32-unknown-elf-g++ (GCC) 9.2.0")
C++ linker for the host machine: /tools/riscv/bin/riscv32-unknown-elf-g++ ld.bfd 2.32
Build machine cpu family: x86_64
Build machine cpu: x86_64
Host machine cpu family: x86_64
Host machine cpu: x86_64
Target machine cpu family: x86_64
Target machine cpu: x86_64
Compiler for C supports arguments -Wa,--no-pad-sections: YES
meson.build:74: WARNING: Consider using the built-in optimization level instead of using "-Os".
meson.build:78: WARNING: Consider using the built-in optimization level instead of using "-Os".
Program python3 found: YES (/usr/bin/python3)
Program objdump found: YES (/usr/bin/objdump)
Program objcopy found: YES (/usr/bin/objcopy)
Program srec_cat found: YES (/usr/bin/srec_cat)
Program git found: YES (/usr/bin/git)
Dependency threads found: YES unknown (cached)
Program cargo found: YES (/home/greg/.cargo/bin/cargo)
Dependency libftdi1 found: YES 1.4 (cached)
Dependency libusb-1.0 found: YES 1.0.21 (cached)
Dependency libcrypto found: YES 1.1.1 (cached)
Build targets in project: 158
```
This repeats over and over

A `meson_init.sh -f` is required to fix it (the `-f` is needed a plain `meson_init.sh` doesn't fix it). I'll see if I can get some steps to reproduce but I think it generally occurs when I've left the build for a while (i.e. yesterday it built successfully as the last thing I did, today I alter a file, run the build and you get the loop) so potentially somehow time dependent indicating something to do with timestamps?

Thought I'd open the issue to see if anyone else has seen similar behaviour or has an idea what may cause it.

**Comment by lenary** [comment]
Ninja will re-run meson when it thinks the generated ninja files are out of date. Presumably something in the meson generation process is making ninja think that the generated ninja files are immediately out of date when generated. Are you running on a nfs or something?

**Comment by GregAC** [comment]
No nfs, just a normal ext4 file system.

**Comment by GregAC** [comment]
Yesterday I took a clean opentitan check and did a new build with `meson_init.sh` followed by `ninja -C ./build-out/`, I then left it. Today I ran `ninja -C ./build-out/` again (with no changes) and I got the infinite loop.

I am wondering if this is actually a meson bug? Possibly something to do with US/UK date mismatch so it ends up getting the same day as a string in DD/MM and MM/DD formats, then when converting them into a comparable datetime they never match, when in fact they do, leading to the infinite rebuilds.

**Comment by lenary** [comment]
I'm not sure if #2596 will help with this, but it might/should (because we no longer edit `build.ninja` without meson/ninja "knowing" about it). How often has this come up again for you?

**Comment by lenary** [comment]
@GregAC has this been an issue for you recently?

**Comment by GregAC** [comment]
Haven't seen it recently but then I've been doing very few software builds. I've grabbed a copy of the latest master and will try and reproduce it with the process I went through in my last comment. I'm on Holiday tomorrow so will let you know how it goes on Monday.

**Comment by GregAC** [comment]
Looks like this is still an issue. Though as no-one else has reported it that means it may be some local setup issue for me rather than an OT issue. If I can find some time I'll take a look at it.

**Comment by kianbeng-lim** [comment]
I have the same problem. The "[0/1] Regenerating build files." repeats over and over if I try to rebuild 
the code after a day or more.

It might be related to the "REGENERATE_BUILD" line in build-out/ninja.build file referencing a file in the "/tmp" folder?

```
build build.ninja: REGENERATE_BUILD ../meson.build ../sw/meson.build ../sw/vendor/meson.build ../sw/device/meson.build ../sw/device/lib/meson.build ../sw/device/lib/base/meson.build ../sw/device/lib/base/freestanding/meson.build ../sw/device/lib/arch/meson.build ../sw/device/lib/runtime/meson.build ../sw/device/lib/testing/meson.build ../sw/device/lib/dif/meson.build ../sw/device/exts/meson.build ../sw/device/exts/common/meson.build ../sw/device/boot_rom/meson.build ../sw/device/boot_rom2/meson.build ../sw/device/examples/meson.build ../sw/device/examples/hello_usbdev/meson.build ../sw/device/examples/hello_world/meson.build ../sw/device/tests/meson.build ../sw/device/tests/dif/meson.build ../sw/device/tests/sim_dv/meson.build ../sw/device/benchmarks/meson.build ../sw/device/benchmarks/coremark/meson.build ../sw/device/tock/meson.build /usr/bin/ln ../sw/device/tock/tock/Cargo_remote.toml ../sw/device/tock/tock/Cargo.toml ../sw/device/riscv_compliance_support/meson.build ../sw/host/meson.build ../sw/host/vendor/meson.build ../sw/host/spiflash/meson.build /tmp/toolchain.yTAepl.txt meson-private/coredata.dat ../meson_options.txt
 pool = console
```
I can trigger the endless loop if I delete that file in the /tmp folder. 
If I restore the file in the /tmp folder the build will complete correctly.

**Comment by imphil** [comment]
I just came to the same conclusion after having the problem on my side, deleting the `/tmp/toolchain.*.txt` file reliably reproduces the problem.

**Comment by lenary** [comment]
Ok, there are two fixes:
- Use the newer toolchains, which don't create a toolchain file in tmp because they ship with a correct one
- The fix in #3167 which avoids using a temporary file, and moves the configuration into the `build-out` directory.
